src/code/padmgr.c

```diff
@@ -23,235 +84,260 @@ OSMesgQueue* PadMgr_LockSerialMesgQueue(PadMgr* padMgr) {
-                osSyncPrintf(VT_FGCOL(YELLOW));
+                PADMGR_LOG(i, "振動パックを認識しました");
+            } else if (ret == PFS_ERR_DEVICE) {
+                padMgr->pakType[i] = CONT_PAK_OTHER;
+            } else if (ret == PFS_ERR_CONTRFAIL) {
+                LOG_NUM("++errcnt", ++sRumbleErrorCount, "../padmgr.c", 282);
+
                 // "Controller pack communication error"
-                osSyncPrintf("padmgr: %dコン: %s\n", i + 1, "コントローラパックの通信エラー");
-                osSyncPrintf(VT_RST);
+                PADMGR_LOG(i, "コントローラパックの通信エラー");
             }
         }
     }
+    sRumbleUpdateCounter++;
 
-    frame++;
-    PadMgr_UnlockSerialMesgQueue(padMgr, serialEventQueue);
+    PadMgr_ReleaseSerialEventQueue(padMgr, serialEventQueue);
 }
 
+/**
+ * Immediately stops rumble on all controllers
+ */
 void PadMgr_RumbleStop(PadMgr* padMgr) {
     s32 i;
-    OSMesgQueue* serialEventQueue = PadMgr_LockSerialMesgQueue(padMgr);
-
-    for (i = 0; i < 4; i++) {
-        if (osMotorInit(serialEventQueue, &padMgr->pfs[i], i) == 0) {
-            if ((gFaultMgr.msgId == 0) && (padMgr->rumbleOnFrames != 0)) {
-                osSyncPrintf(VT_FGCOL(YELLOW));
-                // "Stop vibration pack"
-                osSyncPrintf("padmgr: %dコン: %s\n", i + 1, "振動パック 停止");
-                osSyncPrintf(VT_RST);
-            }
+    OSMesgQueue* serialEventQueue = PadMgr_AcquireSerialEventQueue(padMgr);
+
+    for (i = 0; i < MAXCONTROLLERS; i++) {
+        if (osMotorInit(serialEventQueue, &padMgr->rumblePfs[i], i) == 0) {
+            // If there is a rumble pak attached to this controller, stop it
 
-            osMotorStop(&padMgr->pfs[i]);
+            if (gFaultMgr.msgId == 0 && padMgr->rumbleOnTimer != 0) {
+                // "Stop rumble pak"
+                PADMGR_LOG(i, "振動パック 停止");
+            }
+            osMotorStop(&padMgr->rumblePfs[i]);
         }
     }
 
-    PadMgr_UnlockSerialMesgQueue(padMgr, serialEventQueue);
+    PadMgr_ReleaseSerialEventQueue(padMgr, serialEventQueue);
 }
 
+/**
+ * Prevents rumble for 3 VI, ~0.05 seconds at 60 VI/sec
+ */
 void PadMgr_RumbleReset(PadMgr* padMgr) {
-    padMgr->rumbleOffFrames = 3;
+    padMgr->rumbleOffTimer = 3;
 }
 
-void PadMgr_RumbleSetSingle(PadMgr* padMgr, u32 ctrlr, u32 rumble) {
-    padMgr->rumbleEnable[ctrlr] = rumble;
-    padMgr->rumbleOnFrames = 240;
+/**
+ * Enables or disables rumble on controller port `port` for 240 VI,
+ * ~4 seconds at 60 VI/sec and ~4.8 seconds at 50 VI/sec
+ */
+void PadMgr_RumbleSetSingle(PadMgr* padMgr, u32 port, u32 rumble) {
+    padMgr->rumbleEnable[port] = rumble;
+    padMgr->rumbleOnTimer = 240;
 }
 
-void PadMgr_RumbleSet(PadMgr* padMgr, u8* ctrlrRumbles) {
+/**
+ * Enables or disables rumble on all controller ports for 240 VI,
+ * ~4 seconds at 60 VI/sec and ~4.8 seconds at 50 VI/sec
+ *
+ * @param enable    Array of 4 u8 containing either true or false to enable or disable rumble for that controller
```

Comment:
```suggestion
 * @param enable    Array of u8 containing either true or false to enable or disable rumble for each controller
```
may as well not hardcode the 4 from `MAXCONTROLLERS` in a comment